src/secontext.c

```diff
@@ -161,44 +164,130 @@ selinux_getfilecon(struct tcb *tcp, const char *path, char **secontext,
 	if (!proc_pid)
 		return -1;
 
-	int rc = -1;
-	char fname[PATH_MAX];
+	/*
+	 * The path may contain .../proc/self/... or .../proc/self<eos>.
+	 * We need to replace 'self' by the target PID, or else we would
+	 * resolve the path with 'self' being strace itself.
+	 * In some corner cases, it's impossible to guess so keep it as is if
+	 * we don't find any occurrence of /proc/self.
+	 * Finally there may be breakage if the pathname contains /proc/self
+	 * but has nothing to do with /proc, but it's impossible to tell and
+	 * unlikely to happen.
+	 */
+
+	int rc;
+	char fname[PATH_MAX + 1], *fname_ptr;
+	char buf[PATH_MAX + 1];
+
+	if (strlen(path) >= sizeof(fname))
+		return -1;
+	strcpy(fname, path);
+	bool proc_self_substituted = false;
+
+	/*
+	 * We replace all the occurrences of /proc/self[/] because we cannot
+	 * know how many there are (e.g. /proc/self/root/proc/self/root/foo).
+	 */
+	fname_ptr = fname;
+	for (;;) {
+		char *s = strstr(fname_ptr, "/proc/self");
+		if (s == NULL)
+			break;
+		switch (s[sizeof("/proc/self")-1]) {
+		case '\0':
+			/* path ends with /proc/self */
+			rc = snprintf(buf, sizeof(buf), "/proc/%u", proc_pid);
+			if ((unsigned int) rc >= sizeof(fname) - (s - fname))
+				return -1;
+			strcpy(s, buf);
+			proc_self_substituted = true;
+			goto done_proc_self;
+		case '/':
+			rc = snprintf(buf, sizeof(buf), "/proc/%u/%s",
+				proc_pid, s + sizeof("/proc/self/"));
```

Comment:
`sizeof("/proc/self/")` is off-by-one, should be `strlen("/proc/self/")` or `sizeof("/proc/self/") - 1` or `sizeof("/proc/self")`.